hail/src/main/scala/is/hail/expr/ir/BlockMatrixIR.scala

```diff
@@ -45,18 +62,199 @@ class BlockMatrixLiteral(value: BlockMatrix) extends BlockMatrixIR {
   override protected[ir] def execute(hc: HailContext): BlockMatrix = value
 }
 
-case class BlockMatrixAdd(left: BlockMatrixIR, right: BlockMatrixIR) extends BlockMatrixIR {
+case class BlockMatrixMap2(left: BlockMatrixIR, right: BlockMatrixIR, applyBinOp: ApplyBinaryPrimOp) extends BlockMatrixIR {
```

Comment:
I'm pretty sure IRs that introduce bindings have to be noted in some other files as well. @catoverdrive and @tpoterba do you know where?

Reply:
Why is this an ApplyBinaryPrimOp? What does it bind? Sorry don't have much context.

Reply:
I also think there's probably no place that needs to be updated right now since we don't have a global typecheck/bind analysis, just ones that look at value IR

Reply:
Was planning on changing that to just be an IR. However, it is kind of weird imo that elsewhere ApplyBinaryPrimOp is self contained and is something that can actually be evaluated, while here it is used to just indicate what operation to perform on the two matrices. This also manifests in IRSuite, since Apply*Ops are evaluated to check for equality and that clearly doesn't happen here.

Reply:
should it just be a `BinaryOp` then? e.g. `Add()`, `Divide()`, ...

Reply:
Was considering that, the intention here was to allow for more general functions (potentially with conditionals) that worked element-wise. @danking any additional argument?

Reply:
@tpoterba this is an error, `ApplyBinaryPrimOp` should be `IR`. The way it works is we bind two refs: `"l"` and `"r"`. The `IR` describes how to compute the output element from the two input elements.

Reply:
and so we'll allow any IRs in here? That sounds great, if we can make it fast.

Reply:
Yes. We pattern match to find ones we can send to BLAS. If you write a non-BLASable thing (say an if) then you get the fastest thing we can give you which is nested for loops.

Reply:
and honestly if GCC can't find a vector intrinsic for
```
for i
  for j
    o[i,j] = ... a[i,j] ... b[i,j] ...
```
where the ... is complicated, then I'm not sure we could possibly do anything smarter

Reply:
but if that's what the user wants, why not try to do it anyway.